targettests/Remote-Sim/args_synthesis.cpp

```diff
@@ -0,0 +1,559 @@
+  rx(1.0 * k, q);
+}
+
+void f(cudaq::qubit &q, double x) __qpu__ { rx(x, q); }
+
+void f(cudaq::qubit &q, const std::vector<double> &k) __qpu__ {
+  for (int i = 0; i < k.size(); ++i)
+    rx(k[i], q);
+}
+
+void f(cudaq::qubit &q, const std::vector<float> &k) __qpu__ {
+  for (int i = 0; i < k.size(); ++i)
+    rx(k[i], q);
+}
+
+// FIXME: https://github.com/NVIDIA/cuda-quantum/issues/1383
+// Vector of bool is not fully supported in MLIR mode.
+// void f(cudaq::qubit &q, const std::vector<bool>& k) __qpu__ {
+//   for (int i = 0; i < k.size(); ++i)
+//     if (k[i])
+//         x(q);
+// }
+
+template <typename T, typename = typename std::enable_if<
+                          std::is_integral<T>::value, T>::type>
+void f(cudaq::qubit &q, const std::vector<T> &k) __qpu__ {
+  for (int i = 0; i < k.size(); ++i)
+    rx(1.0 * k[i], q);
+}
+
+// List of all the test functors to run
+static std::vector<std::function<void()>> ALL_TEST_FUNCTORS;
+
+// Main macro to define a test with different signatures.
+// - Dispatch each argument to the test function 'f'.
+// - Add an execution functor to ALL_TEST_FUNCTORS.
+// For example:
+// DEFINE_TEST_KERNEL(test, (int, a, INT_VAL)(std::vector<float>, b,
+// VEC_FLOAT_VAL)(bool, c, BOOL_VAL)(std::vector<int>,
+// d,VEC_INT_VAL)(std::vector<double>, e, VEC_DOUBLE_VAL));
+
+// will be expanded to (define a kernel with a signature and add a test functor
+// to the list)
+
+// void test(int a, std::vector<float> b, bool c, std::vector<int> d,
+//                    std::vector<double> e) __qpu__ {
+//   cudaq::qubit q;
+//   f(q, a);
+//   f(q, b);
+//   f(q, c);
+//   f(q, d);
+//   f(q, e);
+//   mz(q);
+// }
+// const bool addedtest = []() {
+//   ALL_TEST_FUNCTORS.emplace_back([]() {
+//     auto counts = cudaq::sample(test, INT_VAL, VEC_FLOAT_VAL, BOOL_VAL,
+//     VEC_INT_VAL, VEC_DOUBLE_VAL); counts.dump();
+//   });
+//   return true;
+// }();
+
+#define DEFINE_TEST_KERNEL(func_, ...)                                         \
+  void func_(END(PARAMS_LOOP_0 __VA_ARGS__)) __qpu__ {                         \
+    cudaq::qubit q;                                                            \
+    END(VAR_LOOP_0 __VA_ARGS__)                                                \
+    mz(q);                                                                     \
+  }                                                                            \
+  const bool added##func_ = []() {                                             \
+    ALL_TEST_FUNCTORS.emplace_back([]() {                                      \
+      std::cout << "Testing: " << #func_ << "\n";                              \
+      auto counts = cudaq::sample(func_ END(INVOKE_LOOP_0 __VA_ARGS__));       \
+      counts.dump();                                                           \
+    });                                                                        \
+    return true;                                                               \
+  }();
```

Comment:
Nit: semicolon here and semicolon during invocations are produce double-semicolons in the resulting code. There's not really any harm in that, but it could be avoided.
```suggestion
  }()
```